components/calendar/src/chinese_based.rs

```diff
@@ -232,63 +286,133 @@ impl<C: ChineseBased + CalendarArithmetic> ChineseBasedDateInner<C> {
-        );
-        let day = day_i64 as u8;
-        let is_leap_year = Self::new_year_is_leap_year(
-            first_day_of_year,
-            Some(next_solstice),
-            Some(next_new_year),
-        );
-        let leap_month = if is_leap_year {
-            // This doesn't need to be checked for `None`, since `get_leap_month_from_new_year`
-            // will always return a number greater than or equal to 1, and less than 14.
-            NonZeroU8::new(Self::get_leap_month_from_new_year(first_day_of_year))
+    pub(crate) fn chinese_based_date_from_fixed(
+        date: RataDie,
+        iso: Option<Date<Iso>>,
+    ) -> ChineseBasedDateInner<C> {
+        let iso = if let Some(iso_date) = iso {
+            iso_date
         } else {
-            None
+            Iso::iso_from_fixed(date)
         };
-        let cache = ChineseBasedCache {
-            new_year: first_day_of_year,
-            next_new_year,
-            leap_month,
+
+        let epoch_as_iso = Iso::iso_from_fixed(C::EPOCH);
+        let year = iso.year().number - epoch_as_iso.year().number + 1;
+
+        let data_option = C::get_compiled_data_for_year(year);
+        let data_option = if let Some(data) = data_option {
+            if date < data.new_year {
+                C::get_compiled_data_for_year(year - 1)
+            } else if date >= data.next_new_year {
+                C::get_compiled_data_for_year(year + 1)
+            } else {
+                data_option
+            }
+        } else {
+            None
         };
-        // This can use `new_unchecked` because this function is only ever called from functions which
-        // generate the year, month, and day; therefore, there should never be a situation where
-        // creating this ArithmeticDate would fail, since the same algorithms used to generate the ymd
-        // are also used to check for valid ymd.
-        ChineseBasedDateInner(ArithmeticDate::new_unchecked(year, month, day), cache)
+
+        if let Some(data) = data_option {
+            debug_assert!(
+                date < data.next_new_year,
+                "Stored date {date:?} out of bounds!"
+            );
+            let mut cur = data.new_year - 1;
+            let mut month = 0;
+            for length in data.month_lengths {
+                if cur < date {
+                    month += 1;
+                    cur += length as i64;
+                }
+            }
+            debug_assert!(
+                month < 13 || data.leap_month.is_some(),
+                "Cannot have 13 months in a non-leap year!"
+            );
+
+            debug_assert!((0..=12).contains(&(month - 1)), "Month out of bounds!");
+            #[allow(clippy::indexing_slicing)]
+            // Indexing is ok, above assertion checks (month - 1) is in range of array
+            let day_before_month_start = cur - data.month_lengths[(month - 1) as usize] as i64;
```

Comment:
Suggestion: I'd move more of this code into associated functions on `ChineseBasedCompiledData` or `ChineseBasedYearInfo`.